pkg/cloud/ovirt/clients/machineservice.go

```diff
@@ -344,26 +355,70 @@ func  (is *InstanceService) FindVirtualMachineIP(id string,excludeAddr map[strin
 	for _, reportedDevice := range reportedDeviceSlice.Slice() {
 		nicName, _ := reportedDevice.Name()
 		if !nicRegex.MatchString(nicName) {
-			klog.Infof("ovirt vm id: %s ,  skipped nic %s , naming regex mismatch",id, nicName)
+			klog.Infof("ovirt vm id: %s ,  skipped nic %s , naming regex mismatch", id, nicName)
 			continue
 		}
 
 		ips, hasIps := reportedDevice.Ips()
 		if hasIps {
 			for _, ip := range ips.Slice() {
-				ipres, hasAddress := ip.Address()
+				ipAddress, hasAddress := ip.Address()
 
-				if _, ok := excludeAddr[ipres]; ok {
-					klog.Infof("address %s is excluded from usable IPs", ipres)
+				if _, ok := excludeAddr[ipAddress]; ok {
+					klog.Infof("ipAddress %s is excluded from usable IPs", ipAddress)
 					continue
 				}
 
 				if hasAddress {
-					klog.Infof("ovirt vm id: %s , found usable IP %s", id, ipres)
-					return ipres,nil
+					klog.Infof("ovirt vm id: %s , found usable IP %s", id, ipAddress)
+					return ipAddress, nil
 				}
 			}
 		}
 	}
 	return "", fmt.Errorf("coudlnt find usable IP address for vm id: %s", id)
 }
+
+func (is *InstanceService) getAffinityGroups(cID string, agNames []string) (ag []*ovirtsdk.AffinityGroup, err error) {
+	var ags []*ovirtsdk.AffinityGroup
+	res, err := is.Connection.SystemService().ClustersService().
+		ClusterService(cID).AffinityGroupsService().
+		List().Send()
+	if err != nil {
+		return nil, err
+	}
+	agNamesMap := make(map[string]*ovirtsdk.AffinityGroup)
+	for _, af := range res.MustGroups().Slice() {
+		agNamesMap[af.MustName()] = af
+	}
+	for _, agName := range agNames {
+		if _, ok := agNamesMap[agName]; ok {
+			ags = append(ags, agNamesMap[agName])
+		}
+	}
+	return ags, nil
+}
+
+// handleAffinityGroups adds the VM to the provided affinity groups
+func (is *InstanceService) handleAffinityGroups(vm *ovirtsdk.Vm, cID string, agsName []string) error {
+	ags, err := is.getAffinityGroups(cID, agsName)
+	if err != nil {
+		return err
+	}
+	agService := is.Connection.SystemService().ClustersService().
+		ClusterService(cID).AffinityGroupsService()
+	for _, ag := range ags {
+		klog.Infof("Adding machine %v to affinity group %v", vm.MustName(), ag.MustName())
+		_, err = agService.GroupService(ag.MustId()).VmsService().Add().Vm(vm).Send()
+
+		// TODO: bug 1932320: Remove error handling workaround when BZ#1931932 is resolved and backported
+		if err != nil && !errors.As(err, &ovirtsdk.XMLTagNotMatchError{}) {
```

Comment:
`errors.As` should only be used when you actually need the error instance as the specified type, otherwise `errors.Is` should be used.

Reply:
Well here I wanted to check that the err is of the type XMLTagNotMatchError.
If I will use `errors.Is` I will need to have the exact same error, something like:
`!errors.Is(err, ovirtsdk.XMLTagNotMatchError{"action","vm"})`
I have no real objection to it since it is more specific...

Just wondering (for learning purposes) why is `errors.Is` with the specific error should be used instead `error.As` with the type ?

Reply:
Ideally, the match would go to an interface rather than an implementation to use `errors.Is`, as far as I understand.

Reply:
Change to Is with specific error tags